module/zfs/spa.c

```diff
@@ -2316,6 +2320,176 @@ vdev_count_verify_zaps(vdev_t *vd)
+	 * it is after the pivot, and since it is hard to see how two nodes
+	 * could safely boot from the same pool anyway, we do not do MMP
+	 * for root pools.
+	 */
+	if (spa_is_root(spa))
+		return B_FALSE;
+
+	/*
+	 * Skip the test if the current host, and the host who last imported the
+	 * pool, both have (or had) MMP disabled.  If the pool is successfully
+	 * imported, we will then respect the local setting for mmp writes.
+	 */
+	if (ub->ub_mmp_magic == MMP_MAGIC && ub->ub_mmp_delay == 0 &&
+	    zfs_mmp_interval == 0)
+		return (B_FALSE);
+	/*
+	 * If the orig_* values are nonzero, they are the results of an earlier
+	 * tryimport.  If they match the uberblock we just found, then the pool
+	 * has not changed and we return false so we do not test a second time.
+	 */
+	if (orig_txg && orig_timestamp && orig_txg == ub->ub_txg &&
+	    orig_timestamp == ub->ub_timestamp)
+		return (B_FALSE);
+
+	if (hostid == myhostid)
+		return (B_FALSE);
+
+	if (config_pool_state != POOL_STATE_ACTIVE)
+		return (B_FALSE);
+
+	return (B_TRUE);
+}
+
+static int
+spa_activity_check(uberblock_t *ub, spa_t *spa)
+{
+	kcondvar_t mycv;
+	kmutex_t mymtx;
+	hrtime_t duration_factor = 4 * zfs_mmp_fail_intervals;
+	hrtime_t activity_test_duration = duration_factor * zfs_mmp_interval;
+	uint64_t orig_txg = 0;
+	uint64_t orig_timestamp = 0;
+	vdev_t *rvd = spa->spa_root_vdev;
+	int error = 0;
+	nvlist_t *label;
+
+	cv_init(&mycv, NULL, CV_DEFAULT, NULL);
+	mutex_init(&mymtx, NULL, MUTEX_DEFAULT, NULL);
+	mutex_enter(&mymtx);
+
+	orig_txg = ub->ub_txg;
+	orig_timestamp = ub->ub_timestamp;
+	if (ub->ub_mmp_magic == MMP_MAGIC && ub->ub_mmp_delay) {
+		/*
+		 * ub_mmp_delay * vdev_count_leaves() == zfs_mmp_interval
+		 * on the node that last had pool imported.
+		 */
+		activity_test_duration = duration_factor * ub->ub_mmp_delay *
+		    vdev_count_leaves(spa);
+	}
+
+	/* Randomize in case of simultaneous imports */
+	activity_test_duration += spa_get_random(3) *
+	    activity_test_duration / 10;
+
+	activity_test_duration += gethrtime();
+	for (int i = 0; gethrtime() < activity_test_duration ; i++) {
+		if (i > 0) {
+			if (cv_timedwait_sig_hires(&mycv, &mymtx, NANOSEC,
+			    NANOSEC, 0) == 0) {
+				error = EINTR;
+				break;
+			}
+		}
+
+#ifdef _KERNEL
+		printk(
+#else
+		fprintf(stderr,
+#endif
+		    "activity check cycle %d\n", i);
+
+		vdev_uberblock_load(rvd, ub, &label);
+
+		if (orig_txg != ub->ub_txg ||
+		    orig_timestamp != ub->ub_timestamp) {
+			error = EBUSY;
+			break;
+		}
+	}
+
+	mutex_exit(&mymtx);
+	cv_destroy(&mycv);
+
+	/*
+	 * If user canceled import or we detected activity, fail.
+	 */
+
+	if (error == EINTR) {
+		return (SET_ERROR(EINTR));
+	}
+
+	if (error == EBUSY) {
+		cmn_err(CE_WARN, "pool '%s' could not be "
+		    "loaded as it is current active on another"
+		    "system (host: %s hostid: 0x%lx). See: "
+		    "http://zfsonlinux.org/msg/ZFS-8000-EY",
+		    spa_name(spa), "some_hostname", (unsigned long)12121212);
```

Comment:
Let's drop the "some_hostname" and bogus "hostid" if we don't have them handy.  The http://zfsonlinux.org/msg/ZFS-8000-EY is also for a sightly different case, we could add a new entry for this situation.